core/src/main/scala/org/apache/spark/sql/execution/columnar/ColumnTableScan.scala

```diff
@@ -496,8 +496,10 @@ private[sql] final case class ColumnTableScan(
     val statsRow = ctx.freshName("statsRow")
     val deltaStatsRow = ctx.freshName("deltaStatsRow")
     val colNextBytes = ctx.freshName("colNextBytes")
-    val numTableColumns = if (ordinalIdTerm eq null) relationSchema.size
-    else relationSchema.size - ColumnDelta.mutableKeyNames.length // for update/delete
+    val numTableColumns = if (relationSchema.exists(
```

Comment:
Why was the previous one wrong. I mean the ordinalId term check?
What we understood is that perhaps ordinalIDTerm == null meant no upate/delete has happened on it so just take the relationSchema.size else subtract the virtual columns.

Also why did you change size to length? Any issue there?

Please explain how was it wrong earlier?

Reply:
ordinalIDTerm will be set based on output, so if the output projection does not contain the extra columns then this check will go wrong. Specifically since putInto uses the same plan object for join/anti-join, there is a scan there which goes into the joins and is not fed to ColumnUpdateExec directly but still has those additional columns.